### PR TITLE
Refactor: Pytest refactoring and fixes for system test cases

### PR DESCRIPTION
**Pull request by KonikaChaurasiya-GSLab**
## Change Summary

Pytest refactoring

## Related Issue(s)

Fixes #N/A

## Component(s) name

Pytest

## Proposed changes

Define a clear structure to separate:

- Unit test with no CV interaction to test component: `tests/unit`
- System test to run module backend component and which requires CV connection: `tests/system`
- lib to provides a safe place for fixtures, parametrizes, mook data: `tests/lib`

## How to test

```bash
# Go to test folder
$ cd tests/

# Run Pytest
$ pytest --html=report.html --self-contained-html --cov=. --cov-report=html --capture=tee-sys --color yes system
```

### User Checklist

- N/A

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [ ] My change requires a change to the documentation and documentation have been updated accordingly. (check the box if not applicable)
- [ ] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

**Review by guillaumeVilar — COMMENTED** [review]
Thanks a lot for the PR! Please find below my comments. 

1. The Makefile config.py rule did not get updated (still create the config.py for the on-prem CVP). 
We should update it to move away from on-prem IMHO - i.e. remove username, password and just have 2 fields: 
server and token. 

2. The git pre-commit validation failed because an extra line has been added at the end of the file `tests/system/test_cv_device_tools_fqdn.py` 

3. `system/test_cv_container_tools.py` > OK

4. `system/test_cv_device_tools_fqdn.py` > OK

5. `system/test_cv_device_tools_generic.py` > OK

6. `system/test_cv_device_tools_serial.py` > OK

7. `system/test_cv_device_tools.py` > OK but 2 skip. 
As mentionned, we can move one of the device to the undefinied container before running those tests by removing them from the inventory (CVaaS now automatically add the devices to the provisioning page). 
Sample example: 
```
>>> from lib import config
>>> from cvprac.cvp_client import CvpClient
>>> cvp_client.connect(
...     nodes=[config.server],
...     username=config.username,
...     password=config.password,
...     is_cvaas=True,
...     api_token=config.user_token
... )
>>> cvp_client.api.delete_device('50:00:00:cb:38:c2')
{'result': 'success'}
```

8. As a future improvement (maybe in a future PR), we could also add tests for device search by: 
* hostname
* serial number
This is not yet covered by the test suite as of now.

**Comment by KonikaChaurasiya-GSLab** [comment]
@guillaumeVilar : Thanks for the code review. I have updated the PR as per your suggestions. 
For Comment No. 8 As a future improvement : We  already have the tests cases for device search_by hostname and serial number. I have set the self.inventory.search_by = FIELD_HOSTNAME and self.inventory.search_by = FIELD_SERIAL in the respective tests.

File: tests/system/test_cv_device_tools.py [hunk 1]
```diff
@@ -18,14 +18,18 @@
...
 sys.path.append("../")
 sys.path.append("../../")
 from ansible_collections.arista.cvp.plugins.module_utils.device_tools import FIELD_FQDN, FIELD_SYSMAC, FIELD_ID, FIELD_PARENT_NAME, FIELD_PARENT_ID, FIELD_HOSTNAME
 from ansible_collections.arista.cvp.plugins.module_utils.device_tools import DeviceInventory, CvDeviceTools, FIELD_CONTAINER_NAME
 from lib.helpers import time_log
-from lib.utils import cvp_login, get_devices, get_devices_unknown, get_devices_to_move
-from constants_data import ANSIBLE_CV_SEARCH_MODE
+from lib.utils import cvp_login, get_devices, get_devices_unknown, get_devices_to_move, get_cvp_devices_after_move
+from constants_data import ANSIBLE_CV_SEARCH_MODE, CHECK_MODE
+from watchpoints import watch
```
Comment (titom73) [inline tests/system/test_cv_device_tools.py]: [Dev requirement](https://github.com/aristanetworks/ansible-cvp/blob/devel/ansible_collections/arista/cvp/requirements-dev.txt) file should be updated to add this new library

File: ansible_collections/arista/cvp/plugins/module_utils/container_tools.py [hunk 2]
```diff
@@ -848,7 +848,7 @@ def configlets_detach(self, container: str, configlets: List[str]):
         container_info = self.get_container_info(container_name=container)
         detach_configlets = list()
         for configlet in configlets:
-            data = self.__get_configlet_info(configlet_name=configlet['name'])
+            data = self.__get_configlet_info(configlet_name=configlet)
```
Comment (titom73) [inline ansible_collections/arista/cvp/plugins/module_utils/container_tools.py]: Under investigation as it may break execution

**Review by titom73 — CHANGES_REQUESTED** [review]
See comments

File: tests/system/test_cv_device_tools.py (hunk 1 above)
Comment (KonikaChaurasiya-GSLab) [reply tests/system/test_cv_device_tools.py]: @titom73 - This is an unused import . I have added this is for debugging purpose. Will remove it.

File: ansible_collections/arista/cvp/plugins/module_utils/container_tools.py (hunk 2 above)
Comment (KonikaChaurasiya-GSLab) [reply ansible_collections/arista/cvp/plugins/module_utils/container_tools.py]: @titom73 : Yes, this change is breaking the execution. No need of this change. I have back out this change and made changes in system test case accordingly.

**Review by guillaumeVilar — COMMENTED** [review]
Hi, thanks for the PR. System tests now all passed, awesome!

A small detail, the Makefile was not updated with the changes: 
```
AUTH_CONFIG_FILE = lib/config.py

define AUTH_CONFIG
#!/usr/bin/python
# coding: utf-8 -*-

username = "< username >"
password = "< password >"
server = "< cloudvision ip >"
endef
export AUTH_CONFIG
```

Moreover, I tried the other rules and are not triggering the system tests: 
* `make ci` > only run the unit tests
* `make test` > same
* `make test-api` > Give error: 
```
===================================================================================================== no tests ran in 0.88s ======================================================================================================
ERROR: Wrong expression passed to '-m': generic,api: at column 8: unexpected character ","

make: *** [Makefile:32: test-api] Error 4
```

**Review by guillaumeVilar — APPROVED** [review]
LGTM

**Review by titom73 — APPROVED** [review]
LGTM! Awesome work!

**Pull request merged**